pkg/object/udpproxy/session.go

```diff
@@ -0,0 +1,204 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package udpproxy
+
+import (
+	"fmt"
+	"net"
+	"sync/atomic"
+	"time"
+
+	"github.com/megaease/easegress/pkg/logger"
+	"github.com/megaease/easegress/pkg/util/iobufferpool"
+	"github.com/megaease/easegress/pkg/util/timerpool"
+)
+
+type (
+	session struct {
+		upstreamAddr          string
+		downstreamAddr        *net.UDPAddr
+		downstreamIdleTimeout time.Duration
+		upstreamIdleTimeout   time.Duration
+
+		upstreamConn net.Conn
+		writeBuf     chan *iobufferpool.Packet
+		stopChan     chan struct{}
+		stopped      uint32
+	}
+)
+
+func newSession(downstreamAddr *net.UDPAddr, upstreamAddr string, upstreamConn net.Conn,
+	downstreamIdleTimeout, upstreamIdleTimeout time.Duration) *session {
+	s := session{
+		upstreamAddr:          upstreamAddr,
+		downstreamAddr:        downstreamAddr,
+		upstreamConn:          upstreamConn,
+		upstreamIdleTimeout:   upstreamIdleTimeout,
+		downstreamIdleTimeout: downstreamIdleTimeout,
+
+		writeBuf: make(chan *iobufferpool.Packet, 512),
+		stopChan: make(chan struct{}),
+	}
+
+	go func() {
+		var t *time.Timer
+		var idleCheck <-chan time.Time
+
+		if downstreamIdleTimeout > 0 {
+			t = time.NewTimer(downstreamIdleTimeout)
+			idleCheck = t.C
+		}
+
+		for {
+			select {
+			case <-idleCheck:
+				s.Close()
+			case buf, ok := <-s.writeBuf:
+				if !ok {
+					s.Close()
+					continue
+				}
+
+				if t != nil {
+					if !t.Stop() {
+						<-t.C
+					}
+					t.Reset(downstreamIdleTimeout)
+				}
+
+				bufLen := len(buf.Payload)
+				n, err := s.upstreamConn.Write(buf.Bytes())
+				buf.Release()
+
+				if err != nil {
+					logger.Errorf("udp connection flush data to upstream(%s) failed, err: %+v", upstreamAddr, err)
+					s.Close()
+					continue
+				}
+
+				if bufLen != n {
+					logger.Errorf("udp connection flush data to upstream(%s) failed, should write %d but written %d",
+						upstreamAddr, bufLen, n)
+					s.Close()
+				}
+			case <-s.stopChan:
+				if t != nil {
+					t.Stop()
+				}
+				_ = s.upstreamConn.Close()
+				s.cleanWriteBuf()
+				return
+			}
+		}
+	}()
+
+	return &s
+}
+
+// Write send data to buffer channel, wait flush to upstream
+func (s *session) Write(buf *iobufferpool.Packet) error {
+	select {
+	case s.writeBuf <- buf:
+		return nil // try to send data with no check
+	default:
+	}
+
+	var t *time.Timer
+	if s.upstreamIdleTimeout != 0 {
+		t = timerpool.Get(s.upstreamIdleTimeout * time.Millisecond)
+	} else {
+		t = timerpool.Get(60 * time.Second)
+	}
+	defer timerpool.Put(t)
+
+	select {
+	case s.writeBuf <- buf:
+		return nil
+	case <-s.stopChan:
+		buf.Release()
+		return nil
+	case <-t.C:
+		buf.Release()
+		return fmt.Errorf("write data to channel timeout")
+	}
+}
+
+// ListenResponse session listen upstream connection response and send to downstream
+func (s *session) ListenResponse(sendTo *net.UDPConn) {
```

Comment:
if two clients send requests to one server, how to know whom the response is for?